google/cloud/internal/unified_grpc_credentials.h

```diff
@@ -0,0 +1,55 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_UNIFIED_GRPC_CREDENTIALS_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_UNIFIED_GRPC_CREDENTIALS_H
+
+#include "google/cloud/completion_queue.h"
+#include "google/cloud/future.h"
+#include "google/cloud/internal/credentials.h"
+#include "google/cloud/status.h"
+#include "google/cloud/version.h"
+#include <grpcpp/grpcpp.h>
+#include <memory>
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace internal {
+
+class GrpcAuthenticationStrategy {
+ public:
+  virtual ~GrpcAuthenticationStrategy() = default;
+
+  virtual std::shared_ptr<grpc::Channel> CreateChannel(
+      std::string const& endpoint, grpc::ChannelArguments const& arguments) = 0;
+  virtual Status Setup(grpc::ClientContext&) = 0;
```

Comment:
Is `Setup` allowed to be called multiple times on an instance? The name suggests once, but the fact that it's a method like this suggests multiple times.

If it's intended to be called exactly once, would we be able to remove this function from this interface and instead rely on some factory function to "set up" the appropriate subclass exactly once before returning it? Basically, I'm wondering if this function needs to be part of this interface.

But also, I may not be understanding this strategy class (is it an abstract factory?)

Reply:
> Is `Setup` allowed to be called multiple times on an instance? The name suggests once, but the fact that it's a method like this suggests multiple times.

You should call it once per `grpc::ClientContext`, to see this in use:

https://github.com/coryan/google-cloud-cpp/blob/c43945c7190a58deea41adb8933ef77ec625548c/google/cloud/storage/internal/grpc_client.cc#L91-L92

and

https://github.com/coryan/google-cloud-cpp/blob/c43945c7190a58deea41adb8933ef77ec625548c/google/cloud/storage/internal/grpc_client.cc#L130-L131

> If it's intended to be called exactly once, would we be able to remove this function from this interface and instead rely on some factory function to "set up" the appropriate subclass exactly once before returning it? 

What is being set up is the `grpc::ClientContext`.  Maybe `ApplyAuthToContext` or `ConfigureThisContext` are better names?

> But also, I may not be understanding this strategy class (is it an abstract factory?)

I think "Strategy" is the right name, it abstracts two (maybe some day more) implementations of "how to configure gRPC channels and context to have proper authentication".

Reply:
Ahh, then I misunderstood its purpose. A new name sounds good. `ConfigureContext()` sounds pretty good to me. Otherwise, whatever more descriptive name you prefer SGTM.

Reply:
Done.